hexl/thread-pool/thread-pool.hpp

```diff
@@ -0,0 +1,316 @@
+        SetBarrier_Unlocked();  // Wait 'til all jobs are done
+        pool_mutex.unlock();
+      } else {  // Run on single thread
+        pool_mutex.unlock();
+        job(0, 1);
+      }
+    } else {
+      job(0, 1);
+    }
+  }
+
+  // AddRecursiveCalls: Runs a task on next thread available
+  void AddRecursiveCalls(tp_task_t task_a, tp_task_t task_b) {
+    HEXL_CHECK(task_a, "task_a: Require non empty task");
+    HEXL_CHECK(task_b, "task_b: Require non empty task");
+    bool locked = false;
+
+    // Try using thread pool
+    if (!ThreadPool::child) {
+      locked = pool_mutex.try_lock();
+      if (!locked) {
+        task_a(0, 1);
+        task_b(0, 1);
+        return;
+      }
+    }
+
+    if (!setup_done) {
+      SetupThreads_Unlocked(HEXL_NUM_THREADS);  // Setup if thread pool down
+    }
+
+    size_t next = next_thread.fetch_add(2);
+    if (next <= total_threads - 2) {
+      thread_info_t* thread_handler = thread_handlers.at(next++);
+      if (thread_handler->state.load() == static_cast<int>(STATE::DONE)) {
+        thread_handler->task = task_a;
+        thread_handler->state.store(static_cast<int>(STATE::KICK_OFF));
+      } else if (thread_handler->state.load() ==
+                 static_cast<int>(STATE::SLEEPING)) {
+        thread_handler->task = task_a;
+        thread_handler->state.store(static_cast<int>(STATE::KICK_OFF));
+        thread_handler->waker.notify_one();
+      } else {  // In case thread is not on expected state
+        task_a(next - 1, total_threads);
+      }
+
+      thread_handler = thread_handlers.at(next++);
+      if (thread_handler->state.load() == static_cast<int>(STATE::DONE)) {
+        thread_handler->task = task_b;
+        thread_handler->state.store(static_cast<int>(STATE::KICK_OFF));
+      } else if (thread_handler->state.load() ==
+                 static_cast<int>(STATE::SLEEPING)) {
+        thread_handler->task = task_b;
+        thread_handler->state.store(static_cast<int>(STATE::KICK_OFF));
+        thread_handler->waker.notify_one();
+      } else {  // In case thread is not on expected state
+        task_b(next - 1, total_threads);
+      }
+
+      // Implicit barrier
+      for (size_t i = next - 2; i < next; i++) {
+        WaitThread(thread_handlers.at(i));
+      }
+
+      // Next thread to be used
+      next_thread.fetch_add(-2);
+
+    } else {
+      next_thread.fetch_add(-2);
+      task_a(0, 1);
+      task_b(0, 1);
+    }
+
+    if (locked) {
+      pool_mutex.unlock();
+    }
+  }
+
+  // SetupThreads: Spawns new threads if necessary
+  void SetupThreads(size_t n_threads) {
+    std::lock_guard<std::mutex> lock(pool_mutex);
+    SetupThreads_Unlocked(n_threads);
+  }
+
+  // Return threads' handlers
+  std::vector<const thread_info_t*> GetThreadHandlers() {
+    std::lock_guard<std::mutex> lock(pool_mutex);
+    std::vector<const thread_info_t*> handlers;
+    handlers.assign(thread_handlers.begin(), thread_handlers.end());
+    return handlers;
+  }
+};
```

Comment:
This uses RVO and no tmp obj
```suggestion
  std::vector<const thread_info_t*> GetThreadHandlers() const {
    std::lock_guard<std::mutex> lock(pool_mutex);
    // returns a copy
    return std::vector<const thread_info_t*> {thread_handlers.begin(), thread_handlers.end()};
  }
};
```